app/src/main/java/org/mozilla/fenix/gleanplumb/OnDiskMessageMetadataStorage.kt

```diff
@@ -0,0 +1,95 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.gleanplumb
+
+import android.content.Context
+import android.util.AtomicFile
+import androidx.annotation.VisibleForTesting
+import mozilla.components.support.ktx.util.readAndDeserialize
+import mozilla.components.support.ktx.util.writeString
+import org.json.JSONArray
+import org.json.JSONObject
+import java.io.File
+
+internal const val FILE_NAME = "nimbus_messages_metadata.json"
+
+/**
+ * A storage that persists [Message.Metadata] into disk.
+ */
+class OnDiskMessageMetadataStorage(
```

Comment:
This looks like an incomplete CRUD storage layer - where do we delete the metadata? It looks like we're always appending metadata and not considering user storage.

For comparsion with tab state, that is a core part of the browser, so we should avoid losing tab state data where we can (even then, we delete tab data when the tab is removed). With messaging metadata, the remote server can keep sending our users data and we have no way to control this if the _core browser_ is not able to function properly.

Reply:
I remembered @jhugman and I chatted about that, but we ended agreeing that the amount of data that we are saving could take decades, at the rate that we are sending messages to be significant for users. To implement the purging we will need that Nimbus indicates us when messages is not relevant anymore of us to delete. Happy to opened about and re-explore the possibility :)

Reply:
I think we should handle protecting our users if there is a bug a backend service. We do the inverse to avoid DDOS-ing our services from our users.

Reply:
That's a valid point 👍🏽 !